lib/spack/spack/solver/asp.py

```diff
@@ -740,22 +747,29 @@ def package_dependencies_rules(self, pkg, tests):
                                         and pkg.name not in tests):
                         continue
 
-                    if cond == spack.spec.Spec():
+                    # there is a declared dependency of type t
+
+                    # TODO: this ends up being redundant in the output --
+                    # TODO: not sure if we really need it anymore.
+                    # TODO: Look at simplifying the logic in concretize.lp
+                    self.gen.fact(
+                        fn.declared_dependency(dep.pkg.name, dep.spec.name, t))
```

Comment:
I agree this is redundant. I think we can just emit facts like:
```python
fn.dependency_type(dep.pkg.name, cond_id, t)
```
and modify the `concretize.lp` rules such that:
```
depends_on(Package, Dependency, Type)
 :- dependency_conditions_hold(Package, Dependency, ID),
    dependency_type(Package, ID, Type),
    not virtual(Dependency),
    not external(Package).

dependency_conditions_hold(Package, Dependency, ID) :-
  node(Package)
    : dep_cond(Package, Dependency, ID, "node");
      ...
```
I'll try to modify the PR locally to see if the idea works and report under this comment.

Reply:
Confirmed that this idea works. It reduces redundancy, but doesn't give a great speed-up.

Reply:
Fixed by your PR :)